### ISSUE TITLE
How to analyza projects spanning multiple directories, compiled with `-open`

### ISSUE THREAD

**Issue by jberdine** [cristianoc/reanalyze#4]
The attached mini project [open_test.tar.gz](https://github.com/cristianoc/reanalyze/files/4489968/open_test.tar.gz) consists of:
```
$ tree
.
├── bin
│   ├── dune
│   └── test_bin.ml
├── dune-project
├── lib
│   ├── dune
│   └── test_lib.ml
├── test_bin.ml
└── test_lib.opam
```
How should reanalyze be invoked in this case. I think the crux is that there is a library under `lib` that is used in a binary under `bin`, and so they have different directories of object files. I have tried various choices of root directory. E.g. `reanalyze -dce-cmt _build/default/lib/` reports values that are used in the bin as dead, as expected. On the other hand, `reanalyze -dce-cmt _build/default/` does not report anything as dead even though `Test_lib.N.y` is dead.

Additionally, `test_bin` is compiled with `-open Test_lib`, which might complicate things.

**Comment by cristianoc** [comment]
@jberdine to begin with, I find a `test_lib.cmt` but no `test_bin.cmt`.
That by itself is a red flag: we need `.cmt` files for everything that should be take into account.
Any ideas?

**Comment by cristianoc** [comment]
So it looks like using the `dune` trick `dune build @check @all` gets us both `.cmt` files. So that's a start.

Next, here's what the analysis does, in debug mode:
```sh
% Debug=1 ~/reasonml/reanalyze/examples/reanalyze.exe -dce-cmt ./_build/default
  Scanning ./_build/default/bin/.test_bin.eobjs/byte/test_bin.cmt
  addValueReference test_bin.ml:2:11 --> test_lib.ml:12:6
  addValueReference test_bin.ml:2:0 --> pervasives.mli:714:0
  Scanning ./_build/default/lib/.test_lib.objs/byte/test_lib.cmt
  addValueDeclaration x test_lib.ml:5:4
  addValueDeclaration y test_lib.ml:12:6
  addValueDeclaration x test_lib.ml:5:4
  addValueDeclaration y test_lib.ml:12:6
  addValueDeclaration +x test_lib.ml:9:8
  addValueReference test_lib.ml:12:6 --> test_lib.ml:5:4
  addValueReference test_lib.ml:5:4 --> test_lib.ml:9:8
  
File References

  test_lib.ml -->> 
  test_bin.ml -->> test_lib.ml, pervasives.mli
  Live Value test_lib.N.y: 1 references (test_bin.ml:2:11) [0]
  Live Value test_lib.N.M.x: 1 references (test_lib.ml:12:6) [1]
  Live Value test_lib.F.M.+x: 1 references (test_lib.ml:5:4) [0]
```

We can first look at whether this is what you'd expect. And look into reporting separately.

**Comment by cristianoc** [comment]
So it looks like the dependencies are working across files as one would expect.
Then there's he separate question of where files are looked for, for reporting. But I can look into that next.

Basically, the idea is that you would not wan to report on things you don't have the source of.
I'll extend the debug mode a little to show more file paths so one can understand in more detail what's going on.
The starting point of this was the analysis of bucklescript projects, which thankfully use fully qualified source paths so there's no ambiguity.

**Comment by cristianoc** [comment]
Here's the version with full paths:
```
% Debug=1 ~/reasonml/reanalyze/examples/reanalyze.exe -dce-cmt _build/default
  Scanning _build/default/bin/.test_bin.eobjs/byte/test_bin.cmt
  addValueReference bin/test_bin.ml:2:11 --> lib/test_lib.ml:12:6
  addValueReference bin/test_bin.ml:2:0 --> pervasives.mli:714:0
  Scanning _build/default/lib/.test_lib.objs/byte/test_lib.cmt
  addValueDeclaration x lib/test_lib.ml:5:4
  addValueDeclaration y lib/test_lib.ml:12:6
  addValueDeclaration x lib/test_lib.ml:5:4
  addValueDeclaration y lib/test_lib.ml:12:6
  addValueDeclaration +x lib/test_lib.ml:9:8
  addValueReference lib/test_lib.ml:12:6 --> lib/test_lib.ml:5:4
  addValueReference lib/test_lib.ml:5:4 --> lib/test_lib.ml:9:8
  
File References

  test_lib.ml -->> 
  test_bin.ml -->> test_lib.ml, pervasives.mli
  Live Value test_lib.N.y: 1 references (bin/test_bin.ml:2:11) [0]
  Live Value test_lib.N.M.x: 1 references (lib/test_lib.ml:12:6) [1]
  Live Value test_lib.F.M.+x: 1 references (lib/test_lib.ml:5:4) [0]
```

Things look OK to me. Modulo some weird "smartness" of the compiler than somehow thinks the location of `N.y` is kept live, while the actual source code says `N.M.x`.

**Comment by cristianoc** [comment]
@jberdine ball back in your court -- what would you expect to happen on that test?

**Comment by jberdine** [comment]
Ok, I'll try some more to figure out what's going on in the actual code. I guess that the weird way that N.y is considered live threw me. The debug trace helps, thanks.
